### PR TITLE
iestream atomic model to be added to cadmium basic models

### PR DESCRIPTION
**Pull request by cruizm**
This is the iestream model @Laouen and I implemented couple of years ago to get messages from files. I think it is a good idea to have this model inside cadmium basic models as it is used and will probably by almost coupled model.

### CODE REVIEW

File: include/cadmium/basic_model/iestream.hpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+ * Data type MSG must have the operator >> in order to work. 
+ * Data type TIME must also have the operator >> in order to work.
+ * Each line must be an MSG. Therefore the operator >> cannot read inputs that are specified in multiple lines
+ *
+*/
+
+
+template<class TIME, class INPUT>
+class Parser {
+    private:
```
Comment (sdavtaker) [inline include/cadmium/basic_model/iestream.hpp]: Unnecessary.

File: include/cadmium/basic_model/iestream.hpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    public:
+        // Constructors
+        Parser() {}
+        Parser(const char* file_path) {
+            this->open_file(file_path);
+        }
+        void open_file(const char* file_path) {
+            file.open(file_path);
+        }
+    std::pair<TIME,INPUT> next_timed_input() {
```
Comment (sdavtaker) [inline include/cadmium/basic_model/iestream.hpp]: indentation

File: include/cadmium/basic_model/iestream.hpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        iestream_input() noexcept {}
+        iestream_input(const char* file_path) noexcept {
+            state._parser.open_file(file_path);
+        }
+    // state definition
+    struct state_type{
+        Parser<TIME, MSG> _parser;
+        MSG _last_input_read;
+        vector<MSG> _next_input;
+        TIME _simulation_time = TIME();
```
Comment (sdavtaker) [inline include/cadmium/basic_model/iestream.hpp:96]: usually the _ prefix is used for private variables, these here are public

File: include/cadmium/basic_model/iestream.hpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    }; 
+    state_type state;
+
+    // ports definition
+    using input_ports=std::tuple<>;
+    using output_ports=std::tuple<typename defs::out>;
+
+    // internal transition
+    void internal_transition() {
+        //out << "INTERNAL. IESTREAM " << endl;
```
Comment (sdavtaker) [inline include/cadmium/basic_model/iestream.hpp]: You can pass this to the logger to keep track of it. We can add a it to the info level or debug level easily.

File: include/cadmium/basic_model/iestream.hpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                state._next_time = parsed_line.first - state._simulation_time;
+                if (state._next_time < TIME({0})) throw std::exception();
+                state._last_input_read = parsed_line.second;
+                state._next_input.push_back(state._last_input_read);
+            } catch(std::exception& e) {
+                state._next_time = std::numeric_limits<TIME>::infinity();
+            }
+            try {
+                std::pair<TIME, MSG> parsed_line = state._parser.next_timed_input();
+                state._next_time2 = parsed_line.first - state._simulation_time;
```
Comment (sdavtaker) [inline include/cadmium/basic_model/iestream.hpp:124]: can you make the name more explicit, what you mean by next_time2, it is easy to mix up with next_time.

**Review by sdavtaker — COMMENTED** [review]
It would make sense to have a few tests around the model, there is some example of tests on the previous models, and in particular for something like this in the CD++ one.

File: include/cadmium/basic_model/iestream.hpp (hunk 1 above)
Comment (cruizm) [reply include/cadmium/basic_model/iestream.hpp]: Removed the unnecessary private

File: include/cadmium/basic_model/iestream.hpp (hunk 4 above)
Comment (cruizm) [reply include/cadmium/basic_model/iestream.hpp]: It was an old comment for debugging. Removed

File: include/cadmium/basic_model/iestream.hpp (hunk 2 above)
Comment (cruizm) [reply include/cadmium/basic_model/iestream.hpp]: Fixed

File: include/cadmium/basic_model/iestream.hpp (hunk 5 above)
Comment (cruizm) [reply include/cadmium/basic_model/iestream.hpp:124]: next_time2 is the next time after next_time

**Pull request closed without merging**